pallets/gear/src/queue.rs

```diff
@@ -109,120 +115,176 @@ where
+                    log::debug!(
+                        "Re-instrumentation already failed for program '{destination_id:?}' \
+                        with instructions weights version {weights_version}"
+                    );
+
+                    return core_processor::process_code_metadata_error(context);
+                }
+
+                true
+            }
+        };
 
         // Reinstrument the code if necessary.
-        let schedule = T::Schedule::get();
-        let (code, context) =
-            if code.instruction_weights_version() == schedule.instruction_weights.version {
-                (code, ContextChargedForInstrumentation::from(context))
-            } else {
-                log::debug!("Re-instrumenting code for program '{destination_id:?}'");
-
-                let context = match core_processor::precharge_for_instrumentation(
-                    block_config,
-                    context,
-                    code.original_code_len(),
-                ) {
-                    Ok(context) => context,
-                    Err(journal) => return journal,
-                };
+        let (instrumented_code, code_metadata, context) = if needs_reinstrumentation {
+            log::debug!("Re-instrumenting code for program '{destination_id:?}'");
 
-                let code = match Pallet::<T>::reinstrument_code(code_id, &schedule) {
-                    Ok(code) => code,
+            let context = match context
+                .charge_for_original_code(block_config, code_metadata.original_code_len())
+            {
+                Ok(code) => code,
+                Err(journal) => return journal,
+            };
+
+            log::debug!("Gas burned after Original Code {:?}", context.gas_burned());
+
+            let context = match context
+                .charge_for_instrumentation(block_config, code_metadata.original_code_len())
+            {
+                Ok(code) => code,
+                Err(journal) => return journal,
+            };
+
+            log::debug!(
+                "Gas burned after Instrumentation {:?}",
+                context.gas_burned()
+            );
+
+            let instrumented_code_and_metadata =
+                match Pallet::<T>::reinstrument_code(code_id, code_metadata, &schedule) {
+                    Ok(code_and_metadata) => code_and_metadata,
                     Err(e) => {
                         log::debug!("Re-instrumentation error for code {code_id:?}: {e:?}");
                         return core_processor::process_reinstrumentation_error(context);
                     }
                 };
 
-                (code, context)
+            (
+                instrumented_code_and_metadata.instrumented_code,
+                instrumented_code_and_metadata.metadata,
+                context,
+            )
+        } else {
+            // Adjust gas counters for fetching instrumented binary code.
+            let context = match context
```

Comment:
place todo that instrumentation cost should have "write-to-storage" cost inside, so read of instrumented code will be free of charge (since it will come from overlay)